### PR TITLE
feat(client, api): Add Camper Join Date to Profile

### LINKED ISSUE #37817: Add "Joined freeCodeCamp in [month] of [year]" to profile
We'd like to recognize people's original date of account creation on their profile. This is a nice way of recognizing the many early adopters who joined back before freeCodeCamp was a proven path to a developer job, and who were "taking a chance on us" :)

Even though we don't store a date of account creation, it's possible to get a record creation date using the Mongo ID property.

We're looking for a contributor who's interested in helping add "Joined freeCodeCamp in [month] of [year]" to people's portfolios (shown only when they set their bio to public).

### PR DESCRIPTION
**Pull request by Manish-Giri**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Closes #37817

### CODE REVIEW

File: api-server/common/models/user.js
Comment (raisedadead) [inline api-server/common/models/user.js]: This could be done on client instead, and keep the API pure.

**Review by raisedadead — CHANGES_REQUESTED** [review]
I know this is WIP. Left some early comments. Nice work so far. :)

**Review by ahmaxed — APPROVED** [review]

**Review by ahmaxed — CHANGES_REQUESTED** [review]
We could have the joining date as an api object and handle the logic in the client.
For privacy permissions, we could have this handled with" my profile " permissions or "my about me permissions".

File: api-server/common/models/user.js
Comment (Manish-Giri) [reply api-server/common/models/user.js]: Thank You for the review! I'll move the processing logic to the client.

**Comment by Manish-Giri** [comment]
@ahmadabdolsaheb As we discussed, I think all my intended changes are in, so I'm marking it ready for review. @raisedadead I moved the processing logic for (formatting) date to the client. 

1. One thing I observed is - since the development env has only one user, `developmentuser`, when I was testing this change, the `joinDate` prop would show up as `undefined` in the `Camper` component, which in turn, would mean that the actual value of the `joinDate` prop wasn't being rendered correctly to the `Camper` component by the `Profile` component. In the dev env, if you sign-in locally with `developmentuser` (the default login),  there's not an actual portfolio page that is served, instead, you click on the "preview custom profile page" - maybe this is just a static page, that doesn't pull the changes I made to `user.js`, which is why there is no `joinDate` being fetched from `User.findOne()`.  Also, because of this, I couldn't check to make sure if the formatting is correct, when the join date _does_ get displayed successfully. For the time being, I just applied the classes `bio text-center`. 

2. Initially, I was returning a String from `user.js` (as part of `User.findOne()`) itself, since the `parseDate` function would convert the `joinDate`, of type `ISODate`, to a string. Now that the conversion is placed in the client instead, I didn't know what the type of the value would be, when it is destructured in `Profile.js` and passed on as a prop to `Camper.js`, (because of not being able to test it). So, there might be an error with `PropTypes` if the types are different. 

Since you would have the correct access to a DB with some users, I think you would be able to check if the changes are working as intended and let me know if any changes are needed.

**Comment by ahmadabdolsaheb** [comment]
I see. There are two things happening. 

First when users visits their own profile, I believe the client calls '/user/get-session-user'.
If you visit a profile that is not associated with the current session, the client calls '/api/users/get-public-profile?username=${username}'

You can see the server logs for the endpoints that are hit after each UI interaction. 

In order to test the join date with users other than developer user,  set LOCAL_MOCK_AUTH=false in .env and npm run develop after that. Then when you click on the signin button, you will be redirect to our dev auth page. You will be able to create and test multiple accounts if you authenticate with email.

I am not sure if user._id.getTimeStamp() is getting the record creation date. However, you should be able to log user records and see what is being returned at each stage.

Let me know if you get the join date to appear on the UI, or if you have further questions.

File: api-server/common/models/user.js
Comment (ojeytonwilliams) [inline api-server/common/models/user.js:844]:
```suggestion
          joinDate: user.id.getTimeStamp()
```
I'm not sure why it's `id` and not `_id`, but it is.

File: client/src/components/profile/components/Camper.js
Comment (ojeytonwilliams) [inline client/src/components/profile/components/Camper.js:50]:
```suggestion
function parseDate(joinDate) {
  joinDate = new Date(joinDate);
```
I *think* what's going on here is the Date gets serialised to a JSON string.  It looks like we have to recreate the Date manually.  https://weblog.west-wind.com/posts/2014/jan/06/javascript-json-date-parsing-and-real-dates discusses this, and the conclusion seems to be `new Date(dateString)` is fine.

File: client/src/components/profile/components/Camper.js
Comment (ojeytonwilliams) [inline client/src/components/profile/components/Camper.js:64]:
```suggestion
```
There's a more concise way, see below:

File: client/src/components/profile/components/Camper.js
Comment (ojeytonwilliams) [inline client/src/components/profile/components/Camper.js:66]:
```suggestion
  // TODO: revisit this when we localise to different languages.
  const month = joinDate.toLocaleDateString('en-US', { month: 'long' });
```

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
@Manish-Giri the easiest way I found to check this locally is to open an incognito window 
 at http://localhost:8000/developmentuser.  It looks nice!

Regarding the logged in profile, if user a is logged in, then it should show them what user b sees, so it needs `joinDate`.  Could you add `joinDate` to `getSessionUser`?  That should be enough.

File: client/src/components/profile/components/Camper.js
Comment (Manish-Giri) [reply client/src/components/profile/components/Camper.js:66]: @ojeytonwilliams I actually had tried out this way earlier, from here - https://stackoverflow.com/a/18648314

The first code snippet prints `"November"` when you run it in the answer, which is just what we want, but when I tried this with the `ISODate` obtained from MongoDB, the format I got was different - 

Untitled90

I didn't want to spend too much time experimenting with it, before getting the main features done, so I just went ahead with the direct approach [specified in the answer](https://stackoverflow.com/a/1643468).

File: client/src/components/profile/components/Camper.js
Comment (Manish-Giri) [reply client/src/components/profile/components/Camper.js:50]: I haven't yet fixed `getSessionUser()`, but when I tested this in the Mongo Shell, there seems to be an issue during recreating the JSON date with `new Date()` - 

Untitled91

Comment (ojeytonwilliams) [reply client/src/components/profile/components/Camper.js:50]: Sorry for the delay!

So, what's happening here is that `stringify` is inserting quotes around the date string, i.e. `a[0] === '"'`, so `Date` does not understand it. If you parse it, it works:

```js
d = new Date(JSON.parse(a))
// ISODate("2018-10-26T12:52:31Z")
```

File: client/src/components/profile/components/Camper.js
Comment (ojeytonwilliams) [reply client/src/components/profile/components/Camper.js:66]: The Mongo shell doesn't necessarily use the same JavaScript engine as the browser, so the shell's behaviour is likely to not match the browser's.  However, there's reasonable browser support for [toLocaleString](https://caniuse.com/#search=toLocaleString).  What I had not realised that there was better support for that than `toLocaleDateString`, so I updated my suggestion.

I think the support is high enough that we do need a fallback.  @ahmadabdolsaheb what do you think?

Comment (ahmadabdolsaheb) [reply client/src/components/profile/components/Camper.js:66]: > I think the support is high enough that we do need a fallback. @ahmadabdolsaheb what do you think?

Do you mean we would not need a fallback because there is good browser support?

if so, I agree. the iso date should be sent to the client and tested there. If we could get it to work with toLocaleString, I think we should go with it.

Comment (ojeytonwilliams) [reply client/src/components/profile/components/Camper.js:66]: Yep. Caniuse has 95% support and I assume (maybe you have stats on this) that most people using fcc have reasonably up-to-date browsers.

**Comment by raisedadead** [comment]
> In order to test the join date with users other than developer user, set LOCAL_MOCK_AUTH=false in .env and npm run develop after that.

Additionally you would need to signup for a auth0 instance (its free), create an application and setup the client id and credentials.

File: client/src/components/profile/components/Camper.js
Comment (raisedadead) [reply client/src/components/profile/components/Camper.js:66]: Technically we support only the latest 'LTS' Operating Systems and browsers.

**Comment by ahmadabdolsaheb** [comment]
how is this coming along @Manish-Giri? cannot wait to see it on production.

**Comment by moT01** [comment]
Looks like one of the tests got stuck or something. Closing/opening to try and run that again.

File: api-server/common/models/user.js
Comment (ahmadabdolsaheb) [reply api-server/common/models/user.js:844]: I get `user.id.getTimeStamp is not a function` for id
@Manish-Giri, it would be great if you could look into this.

Comment (Manish-Giri) [reply api-server/common/models/user.js:844]: @ahmadabdolsaheb I was able to finally sign in with a local user using Auth0, looking into this now.

Comment (Manish-Giri) [reply api-server/common/models/user.js:844]: @ojeytonwilliams Adding `joinDate` to `getSessionUser()` worked, many thanks!

**Comment by Manish-Giri** [comment]
@ahmadabdolsaheb @ojeytonwilliams So finally, I got the join date to appear on the UI, screenshot -

Screenshot 2020-06-02 02 37 44

While I clean up the code and run some tests (privacy of join date on timeline, browser support, etc), if you have any suggestions for the UI in the meantime, pls let me know.

I personally felt the gap between the join date and the number of points felt a bit jarring

https://github.com/freeCodeCamp/freeCodeCamp/blob/31614fa9a6632a3ac86530c546d0a9c8c51d5b55/client/src/components/profile/components/Camper.js#L122

This was after I removed the `br` -

Screenshot 2020-06-02 02 42 58

**Comment by ojeytonwilliams** [comment]
Nice work, @Manish-Giri!  As for the UI, I feel like the screenshot without the break feels a little monolithic.  Maybe a faint horizontal line would help break things up.

**Comment by Manish-Giri** [comment]
@ojeytonwilliams Thanks for the suggestion! I replaced the `br` with `hr`. 
@ahmadabdolsaheb It is ready for review now.

A few points - 

1. Privacy for the join date is controlled by this "about me" option in the settings -

Screenshot 2020-06-02 23 11 38 I tested the visibility with the option turned on and off.

2. The join date is now visible (if "about me" is turned on) for both the session (logged-in) user, and if a user X visits the profile for another user Y, here are the screenshots for my test user and then "developmentuser" -

Screenshot 2020-06-02 22 55 46

Screenshot 2020-06-02 22 54 57

3. The build failure is probably due to the snapshot change caused by the new line introduced by the `hr` tag.

**Comment by ahmadabdolsaheb** [comment]
Great job, @Manish-Giri,

Have you pushed your latest work? It seems like the last commit was on Mar 21.
Ideally, we could go in the direction of twitter profile where we divide the data into groups and display them rows.

I could go in and do that after this pr is merged.

For now, I believe we could go with "{calendar icon} Joined {month} {year}" format to display the timestamp so we could use the vertical space for other items.

**Comment by Manish-Giri** [comment]
@ahmadabdolsaheb Thank You for the review. Yes, I rebased my PR with the latest `master` before submitting, my latest commits were https://github.com/freeCodeCamp/freeCodeCamp/pull/38408/commits/57d7b4e3be2cf26cc0f7c8d0361ed5db567cbee0 and https://github.com/freeCodeCamp/freeCodeCamp/pull/38408/commits/68e8b4dc1ef59ccff1c36694c68d39c156675d63, which were just a few hours ago. I think the other commits from March are also listed because I force-pushed the branch, to bring in the latest changes from `master`.

Regarding the calendar icon, would you happen to know which package/library it is from (such as Font Awesome)? If so, I could go in and check the usage and then re-submit with the changes.

**Comment by ojeytonwilliams** [comment]
You can see from
image

that the two `<hr>`s differ in length.  I think the problem is that part of the timeline does not have the same bootstrap row/column structure as the rest.

**Comment by Manish-Giri** [comment]
Yes, if I remember correctly, the second `hr` is in a `div` which has the `col-xs-8` class, thereby making the line length shorter, while the first `hr` is just in a regular `div`.

**Comment by ahmadabdolsaheb** [comment]
> Regarding the calendar icon, would you happen to know which package/library it is from (such as Font Awesome)? If so, I could go in and check the usage and then re-submit with the changes.

on camper.js the following two icons are used.
`import { faAward, faHeart } from '@fortawesome/free-solid-svg-icons';`

you could find a relevant one for calendar.

**Comment by Manish-Giri** [comment]
@ahmadabdolsaheb Ready for review.

**Review by ahmaxed — APPROVED** [review]

**Comment by ahmadabdolsaheb** [comment]
Great job @Manish-Giri,
It would be great if you could add a cypress test. just to lock this feature down and provide a test example in the code codebase for other developers developing similar features.

**Comment by ahmadabdolsaheb** [comment]
@ojeytonwilliams, I think we could merge this without the tests if it LGTY.

**Comment by Manish-Giri** [comment]
@ahmadabdolsaheb I have time to work on this again. I can add some tests if you still prefer.

I haven't used Cypress before, but going by the docs, it shouldn't be too hard.

**Review by raisedadead — APPROVED** [review]

**Comment by raisedadead** [comment]
Hi Manish, thanks for your hard work on this. 

I would encourage you to keep working on the e2e tests or unit tests if you can. That would solidify the implementation.

Happy contributing.

**Comment by Manish-Giri** [comment]
@raisedadead Thanks for the merge. I was finalizing my graduation requirements last few weeks, so got a little busy. I'll start working on the e2e implementation now.

**Comment by raisedadead** [comment]
Awesome, goodluck with both of your graduation and e2e test 😃 !

**Pull request merged**
